File: sdks/java/src/main/java/do_/kafka/RecordMetadata.java
package do_.kafka;

import java.time.Instant;
import java.util.Objects;

/**
 * Metadata about a record that has been acknowledged by the server.
 */
public final class RecordMetadata {
    private final String topic;
    private final int partition;
    private final long offset;
    private final Instant timestamp;
    private final int serializedKeySize;
    private final int serializedValueSize;

    /**
     * Creates record metadata.
     *
     * @param topic               The topic the record was appended to
     * @param partition           The partition the record was sent to
     * @param offset              The offset of the record in the topic/partition
     * @param timestamp           The timestamp of the record
     * @param serializedKeySize   The size of the serialized key
     * @param serializedValueSize The size of the serialized value
     */
    public RecordMetadata(String topic, int partition, long offset, Instant timestamp,
                          int serializedKeySize, int serializedValueSize) {
        this.topic = topic;
        this.partition = partition;
        this.offset = offset;
        this.timestamp = timestamp;
        this.serializedKeySize = serializedKeySize;
        this.serializedValueSize = serializedValueSize;
    }

    /**
     * Creates record metadata with minimal fields.
     */
    public RecordMetadata(String topic, int partition, long offset) {
        this(topic, partition, offset, Instant.now(), -1, -1);
    }

    /**
     * Returns the topic the record was appended to.
     */
    public String topic() {
        return topic;
    }

    /**
     * Returns the partition the record was sent to.
     */
    public int partition() {
        return partition;
    }

    /**
     * Returns the offset of the record in the topic/partition.
     */
    public long offset() {
        return offset;
    }

    /**
     * Returns the timestamp of the record.
     */
    public Instant timestamp() {
        return timestamp;
    }

    /**
     * Returns the size of the serialized key, or -1 if not available.
     */
    public int serializedKeySize() {
        return serializedKeySize;
    }

    /**
     * Returns the size of the serialized value, or -1 if not available.
     */
    public int serializedValueSize() {
        return serializedValueSize;
    }

    /**
     * Returns the TopicPartition for this record.
     */
    public TopicPartition topicPartition() {
        return new TopicPartition(topic, partition);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof RecordMetadata that)) return false;
        return partition == that.partition &&
                offset == that.offset &&
                Objects.equals(topic, that.topic);
    }

    @Override
    public int hashCode() {
        return Objects.hash(topic, partition, offset);
    }

    @Override
    public String toString() {
        return "RecordMetadata{" +
                "topic='" + topic + '\'' +
                ", partition=" + partition +
                ", offset=" + offset +
                ", timestamp=" + timestamp +
                '}';
    }
}

File: sdks/ruby/lib/kafka_do/record_metadata.rb
# frozen_string_literal: true

require 'time'

module KafkaDo
  # Metadata about a successfully produced record
  #
  # @example Accessing record metadata
  #   metadata = producer.produce('Hello!', topic: 'my-topic')
  #   puts "Sent to partition #{metadata.partition} at offset #{metadata.offset}"
  #
  class RecordMetadata
    # @return [String] the topic the record was sent to
    attr_reader :topic

    # @return [Integer] the partition the record was sent to
    attr_reader :partition

    # @return [Integer] the offset of the record in the partition
    attr_reader :offset

    # @return [Time] the timestamp of the record
    attr_reader :timestamp

    # @return [Integer] the size of the serialized key
    attr_reader :serialized_key_size

    # @return [Integer] the size of the serialized value
    attr_reader :serialized_value_size

    # Create a new RecordMetadata
    #
    # @param topic [String] the topic name
    # @param partition [Integer] the partition number
    # @param offset [Integer] the offset
    # @param timestamp [Time] the timestamp
    # @param serialized_key_size [Integer] serialized key size
    # @param serialized_value_size [Integer] serialized value size
    def initialize(topic:, partition:, offset:, timestamp: nil,
                   serialized_key_size: 0, serialized_value_size: 0)
      @topic = topic
      @partition = partition
      @offset = offset
      @timestamp = timestamp || Time.now
      @serialized_key_size = serialized_key_size
      @serialized_value_size = serialized_value_size
    end

    # Create RecordMetadata from an RPC response hash
    #
    # @param data [Hash] the RPC response data
    # @return [RecordMetadata] the parsed metadata
    def self.from_rpc(data)
      timestamp = if data['timestamp']
                    Time.at(data['timestamp'] / 1000.0)
                  else
                    Time.now
                  end

      new(
        topic: data['topic'],
        partition: data['partition'].to_i,
        offset: data['offset'].to_i,
        timestamp: timestamp,
        serialized_key_size: data['serialized_key_size'].to_i,
        serialized_value_size: data['serialized_value_size'].to_i
      )
    end

    # Returns the TopicPartition for this record
    #
    # @return [TopicPartition]
    def topic_partition
      TopicPartition.new(topic, partition)
    end

    # Returns a string representation
    #
    # @return [String]
    def to_s
      "#<KafkaDo::RecordMetadata topic=#{topic} partition=#{partition} offset=#{offset}>"
    end

    # Returns a detailed inspection string
    #
    # @return [String]
    def inspect
      "#<KafkaDo::RecordMetadata topic=#{topic.inspect} partition=#{partition} " \
        "offset=#{offset} timestamp=#{timestamp}>"
    end
  end
end
